libs/llrb/llrb.go

```diff
@@ -0,0 +1,256 @@
+package llrb
+
+import (
+	"fmt"
+	"math"
+	"sync"
+)
+
+// MaxSize is the max allowed number of node a llrb is allowed to contain
+const MaxSize = int(^uint(0) >> 1)
+
+func (a NodeKey) compare(b NodeKey) int {
+	if a.Priority > b.Priority {
+		return 1
+	}
+	if a.Priority < b.Priority {
+		return -1
+	}
+	if a.TS.Before(b.TS) {
+		return 1
+	}
+	if a.TS.After(b.TS) {
+		return -1
+	}
+	return 0
+}
+
+type node struct {
+	key         NodeKey
+	data        interface{}
+	left, right *node
+	black       bool
+}
+
+type llrb struct {
+	mtx     sync.RWMutex
+	size    int
+	maxSize int
+	root    *node
+}
+
+// newLLRB return llrb with given maxSize
+func newLLRB(maxSize int) *llrb {
+	return &llrb{maxSize: maxSize}
+}
+
+// Size returns the number of nodes in the tree
+func (t *llrb) Size() int {
+	t.mtx.RLock()
+	defer t.mtx.RUnlock()
+	return t.size
+}
+
+// GetNext retrieves a satisfied tx with "largest" nodeKey and "smaller" than starter if provided
+func (t *llrb) GetNext(starter *NodeKey, predicate func(interface{}) bool) (interface{}, error) {
+	t.mtx.RLock()
+	defer t.mtx.RUnlock()
+	startKey := NodeKey{
+		Priority: uint64(math.MaxUint64),
+	}
+	if starter != nil {
+		startKey = *starter
+	}
+	var candidate *node
+	for h := t.root; h != nil; {
+		if h.key.compare(startKey) == -1 && predicate(h.data) {
+			if candidate == nil || candidate.key.compare(h.key) == -1 {
+				candidate = h
+			}
+			h = h.right
+		} else {
+			h = h.left
+		}
+	}
+	if candidate == nil {
+		return nil, nil
```

Comment:
as said before, returning nil to call site can cause troubles sometimes (see [this](https://medium.com/@mangatmodi/go-check-nil-interface-the-right-way-d142776edef1)). how about returning a specific type of error?

```
// llrb.go
var ErrorStopIteration = errors.New(“STOP ITERATION”)
...
return nil, ErrorStopIteration

// mempool.go
data, err := llrb.GetNext()
if err == llrb.ErrStopIteration {...}
else if err != nil {...}
else ...
```